iroha_p2p/src/network.rs

```diff
@@ -0,0 +1,195 @@
+use std::{collections::HashMap, io};
+
+use async_stream::stream;
+use futures::Stream;
+use iroha_actor::{broker::Broker, Actor, Addr, Context, ContextHandler, Handler};
+use iroha_logger::warn;
+use parity_scale_codec::{Decode, Encode};
+use tokio::net::{TcpListener, TcpStream};
+use ursa::{encryption::symm::Encryptor, kex::KeyExchangeScheme};
+
+use crate::peer::{Peer, PeerId, State};
+
+/// Main network layer structure, that is holding connections, called [`Peer`]s.
+#[derive(Debug)]
+pub struct Network<T, K, E>
+where
+    T: Encode + Decode + Send + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    peers: HashMap<PeerId, Addr<Peer<T, K, E>>>,
+    listener: Option<TcpListener>,
+    broker: Broker,
+    //handler: AlwaysAddr<A>
+}
+
+impl<T, K, E> Network<T, K, E>
+where
+    T: Encode + Decode + Send + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    /// Creates a network structure, that will hold connections to other nodes.
+    pub fn new(broker: Broker /*handler: AlwaysAddr<A>*/) -> Self {
+        Self {
+            peers: HashMap::new(),
+            listener: None,
+            broker,
+            /*handler*/
+        }
+    }
+
+    fn listener_stream(&mut self) -> impl Stream<Item = NewPeer> + Send + 'static {
+        #[allow(clippy::unwrap_used)]
+        let listener = self.listener.take().unwrap();
+        stream! {
+            loop {
+                match listener.accept().await {
+                    Ok((stream, addr)) => {
+                        let id = PeerId { address: addr.to_string(), public_key: None };
+                        let new_peer: NewPeer = NewPeer(Ok((stream, id)));
+                        yield new_peer;
+                    },
+                    Err(error) => {
+                        yield NewPeer(Err(error));
+                        break;
```

Comment:
Should we break here btw? I mean there might be case that we failed to accept one malicious tcp client, but generally our connection is ok